From 637444067e21f2de24f60bd9190fc4ab92e1fcba Mon Sep 17 00:00:00 2001
From: Todd Lipcon <todd@cloudera.com>
Date: Mon, 23 Jul 2012 08:54:54 -0700
Subject: [PATCH 107/795] MR1: MAPREDUCE-4399. Performance regression in shuffle

Reason: performance bugfix
Ref: CDH-6700
Author: Luke Lu
(cherry picked from commit 91f8a876538c1ad425ef6ed2433b1e1aa6b6291b)
(cherry picked from commit 60cb07fa4dfa5fa76e5f027f5988497ecbdd7bb1)
---
 .../org/apache/hadoop/mapred/TaskTracker.java      |   10 +++++-----
 1 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/src/mapred/org/apache/hadoop/mapred/TaskTracker.java b/src/mapred/org/apache/hadoop/mapred/TaskTracker.java
index cb875f4..9218348 100644
--- a/src/mapred/org/apache/hadoop/mapred/TaskTracker.java
+++ b/src/mapred/org/apache/hadoop/mapred/TaskTracker.java
@@ -3969,7 +3969,8 @@ public synchronized void clear() {
   public static class MapOutputServlet extends HttpServlet {
     private static final long serialVersionUID = 1L;
     private static final int MAX_BYTES_TO_READ = 64 * 1024;
-    
+    // work around jetty internal buffering issues
+    private static final int RESPONSE_BUFFER_SIZE = MAX_BYTES_TO_READ + 16;
     private static LRUCache<String, Path> fileCache = new LRUCache<String, Path>(FILE_CACHE_SIZE);
     private static LRUCache<String, Path> fileIndexCache = new LRUCache<String, Path>(FILE_CACHE_SIZE);
     
@@ -4073,10 +4074,9 @@ public void doGet(HttpServletRequest request,
         //set the custom "for-reduce-task" http header to the reduce task number
         //for which this map output is being transferred
         response.setHeader(FOR_REDUCE_TASK, Integer.toString(reduce));
-        
-        //use the same buffersize as used for reading the data from disk
-        response.setBufferSize(MAX_BYTES_TO_READ);
-        
+
+        response.setBufferSize(RESPONSE_BUFFER_SIZE);
+
         /**
          * Read the data from the sigle map-output file and
          * send it to the reducer.
-- 
1.7.0.4

